cdc/log.cc

```diff
@@ -675,15 +675,102 @@ utils::UUID generate_timeuuid(api::timestamp_type t) {
 }
 
+class log_mutation_builder {
+    const schema& _base_schema;
+    const schema& _log_schema;
+    const column_definition& _op_col;
+    const column_definition& _ttl_col;
+
+    // The base mutation's partition key
+    const partition_key& _base_pk;
+
+    // The cdc$time value of created rows
+    const bytes _tuuid;
+    // The timestamp of the created log mutation cells
+    const api::timestamp_type _ts;
+    // The ttl of the created log mutation cells
+    const ttl_opt _ttl;
+
+    // Keeps the next cdc$batch_seq_no value
+    int _batch_no = 0;
+
+    // The mutation under construction
+    mutation& _log_mut;
+
+public:
+    log_mutation_builder(mutation& log_mut, api::timestamp_type ts,
+                         const partition_key& base_pk, const schema& base_schema)
+        : _base_schema(base_schema), _log_schema(*log_mut.schema()),
+          _op_col(*_log_schema.get_column_definition(log_meta_column_name_bytes("operation"))),
+          _ttl_col(*_log_schema.get_column_definition(log_meta_column_name_bytes("ttl"))),
+          _base_pk(base_pk),
+          _tuuid(timeuuid_type->decompose(generate_timeuuid(ts))),
+          _ts(ts),
+          _ttl(_base_schema.cdc_options().ttl()
+                  ? std::optional{std::chrono::seconds(_base_schema.cdc_options().ttl())} : std::nullopt),
+          _log_mut(log_mut)
+    {}
+
+    clustering_key allocate_new_log_row() {
+        const auto log_ck = clustering_key::from_exploded(_log_schema, { _tuuid, int32_type->decompose(_batch_no++) });
+        auto pk_value = _base_pk.explode(_base_schema);
+        size_t pos = 0;
+        for (const auto& column : _base_schema.partition_key_columns()) {
+            assert (pos < pk_value.size());
+
+            auto& cdef = *_log_schema.get_column_definition(log_data_column_name_bytes(column.name()));
+            _log_mut.set_cell(log_ck, cdef, atomic_cell::make_live(*column.type, _ts, bytes_view(pk_value[pos]), _ttl));
+            ++pos;
+        }
+        return log_ck;
+    }
+
+    void set_clustering_columns(const clustering_key& log_ck, const clustering_key_prefix& base_ckey) {
+        auto ck_value = base_ckey.explode(_log_schema);
+        size_t pos = 0;
+        for (const auto& column : _base_schema.clustering_key_columns()) {
+            if (pos >= ck_value.size()) {
+                break;
+            }
+            auto& cdef = *_log_schema.get_column_definition(log_data_column_name_bytes(column.name()));
+            _log_mut.set_cell(log_ck, cdef, atomic_cell::make_live(*column.type, _ts, bytes_view(ck_value[pos]), _ttl));
+            ++pos;
+        }
+    }
+
+    void set_operation(const clustering_key& log_ck, operation op) {
+        _log_mut.set_cell(log_ck, _op_col, atomic_cell::make_live(
+                    *_op_col.type, _ts, _op_col.type->decompose(operation_native_type(op)), _ttl));
+    }
+
+    void set_ttl(const clustering_key& log_ck, ttl_opt ttl) {
+        if (ttl) {
```

Comment:
I think this should be `opt_set_ttl` and in turn call an unconditional `set_ttl` (non-opt arg). It is weird with setters that are internally optional imho.

Reply:
I think of it as follows:
- if `ttl` is non-null, we set `cdc$ttl` to its value
- if it's null, we set `cdc$ttl` to `null`, and the fact that the setter doesn't need to do anything is an implementation detail.

Reply:
To make it completely kosher the setter should actually somehow set it to `null` if the passed ttl is null (e.g. by removing a cell if it exists). Because what if we call `set_ttl` with non-null and then want to revert? But meh.

Reply:
Then please comment on the logic so the person needing to explicitly null something knows the if is because of said assumption.

Reply:
I documented the methods, and included a comment about calling `set_ttl` with null/non-null.